File: mobile/lib/features/deteksi_dini/domain/calculator_anjuran_resolver.dart
import '../models/calculator_anjuran_rule.dart';

class CalculatorAnjuranResolver {
  const CalculatorAnjuranResolver();

  ResolvedAnjuran? resolve({
    required List<CalculatorAnjuranRule> rules,
    required String metric,
    required double value,
    String? indicator,
  }) {
    if (rules.isEmpty) {
      return null;
    }

    final filtered = rules
        .where((rule) => rule.metric == metric)
        .where(
          (rule) => indicator == null
              ? rule.indicator == null
              : rule.indicator == indicator,
        )
        .toList()
      ..sort((a, b) => a.sortOrder.compareTo(b.sortOrder));

    if (filtered.isEmpty) {
      return null;
    }

    CalculatorAnjuranRule? defaultRule;

    for (final rule in filtered) {
      if (rule.isDefault) {
        defaultRule = rule;
        continue;
      }

      if (_matches(rule, value)) {
        return _fromRule(rule);
      }
    }

    if (defaultRule != null) {
      return _fromRule(defaultRule);
    }

    return null;
  }

  bool _matches(CalculatorAnjuranRule rule, double value) {
    final threshold = rule.threshold;
    if (threshold == null) {
      return false;
    }

    switch (rule.operator) {
      case CalculatorAnjuranRule.operatorGt:
        return value > threshold;
      case CalculatorAnjuranRule.operatorGte:
        return value >= threshold;
      case CalculatorAnjuranRule.operatorLt:
        return value < threshold;
      case CalculatorAnjuranRule.operatorLte:
        return value <= threshold;
      default:
        return false;
    }
  }

  ResolvedAnjuran _fromRule(CalculatorAnjuranRule rule) {
    return ResolvedAnjuran(
      slug: rule.slug?.trim().isNotEmpty == true ? rule.slug!.trim() : 'unknown',
      label: rule.label,
      anjuran: rule.anjuran,
    );
  }
}

File: backend/app/Services/Screening/CalculatorAnjuranResolver.php
<?php

namespace App\Services\Screening;

use App\Models\CalculatorAnjuranRule;
use App\Models\EducationContent;
use RuntimeException;

class CalculatorAnjuranResolver
{
	public function resolve(
		EducationContent $content,
		string $metric,
		float $value,
		?string $indicator = null,
	): ResolvedAnjuran {
		$rules = $content->anjuranRules()
			->where('metric', $metric)
			->when(
				$indicator !== null,
				fn ($query) => $query->where('indicator', $indicator),
				fn ($query) => $query->whereNull('indicator'),
			)
			->orderBy('sort_order')
			->get();

		if ($rules->isEmpty()) {
			throw new RuntimeException('Aturan anjuran belum dikonfigurasi untuk konten ini.');
		}

		$defaultRule = null;

		foreach ($rules as $rule) {
			if ($rule->is_default) {
				$defaultRule = $rule;

				continue;
			}

			if ($this->matches($rule, $value)) {
				return $this->fromRule($rule);
			}
		}

		if ($defaultRule !== null) {
			return $this->fromRule($defaultRule);
		}

		throw new RuntimeException('Tidak ada aturan anjuran yang cocok untuk nilai ini.');
	}

	private function matches(CalculatorAnjuranRule $rule, float $value): bool
	{
		if ($rule->threshold === null) {
			return false;
		}

		return match ($rule->operator) {
			CalculatorAnjuranRule::OPERATOR_GT => $value > $rule->threshold,
			CalculatorAnjuranRule::OPERATOR_GTE => $value >= $rule->threshold,
			CalculatorAnjuranRule::OPERATOR_LT => $value < $rule->threshold,
			CalculatorAnjuranRule::OPERATOR_LTE => $value <= $rule->threshold,
			default => false,
		};
	}

	private function fromRule(CalculatorAnjuranRule $rule): ResolvedAnjuran
	{
		return new ResolvedAnjuran(
			slug: (string) ($rule->slug ?: 'unknown'),
			label: $rule->label,
			anjuran: $rule->anjuran,
		);
	}
}
